openedx_tagging/core/tagging/import_export/dsl.py

```diff
@@ -0,0 +1,182 @@
+        Add to the action list and the indexed actions
+        """
+        action = action_cls(self.taxonomy, tag, len(self.actions))
+
+        # We validate if there are no inconsistencies when executing this action
+        self.errors.extend(action.validate(self.indexed_actions))
+
+        # Add action
+        self.actions.append(action)
+
+        # Index the actions for search
+        self.indexed_actions[action.name].append(action)
+
+    def _build_delete_actions(self, tags: dict):
+        """
+        Adds delete actions for `tags`
+        """
+        for tag in tags.values():
+            for child in tag.children.all():
+                if child.external_id not in tags:
+                    # If the child is not to be removed, 
+                    # then update its parent
+                    self._build_action(
+                        UpdateParentTag,
+                        TagDSL(
+                            id=child.external_id,
+                            value=child.value,
+                            parent_id=None,
+                        )
+                    )
+
+            # Delete action
+            self._build_action(
+                DeleteTag,
+                TagDSL(
+                    id=tag.external_id,
+                    value=tag.value,
+
+                )
+            )
+
+    def generate_actions(
+        self,
+        tags: List[TagDSL],
+        replace=False,
+    ):
+        """
+        Generates actions from `tags`.
+        
+        Validates each action and create respective errors
+        If `replace` is True, then deletes the tags that have not been read
+
+        TODO: Missing join/reduce actions. Ex. A tag may have no changes, 
+        but then its parent needs to be updated because its parent is deleted.
+        Those two actions should be merged.
+        """
+        self.actions.clear()
+        self.errors.clear()
+        self._init_indexed_actions()
+        tags_for_delete = {}
+
+        if replace:
+            tags_for_delete = {
+                tag.external_id: tag
+                for tag in self.taxonomy.tag_set.all()
+            }
+
+        for tag in tags:
+            has_action = False
+
+            # Check all available actions and add which ones should be executed
+            for action_cls in available_actions:
+                if action_cls.valid_for(self.taxonomy, tag):
+                    self._build_action(action_cls, tag)
+                    has_action = True
+            
+            if not has_action:
+                # If it doesn't find an action, a "without changes" is added
+                self._build_action(WithoutChanges, tag)
+
+            if replace:
+                tags_for_delete.pop(tag.id)
+
+        if replace:
+            # Delete all not readed tags
+            self._build_delete_actions(tags_for_delete)
+
+    def execute(self) -> List[ImportError]:
+        pass
+
+    def plan(self) -> str:
```

Comment:
Nice.. I like how your "action" classes make this possible.

Reply:
Thanks :smile: